### PR TITLE
PCI

### PR DESCRIPTION
**Pull request by SpaceboyRoss01**
Adding PCI, currently it doesn't do much but I plan to add more code I just need to know how the devs want it to work.

### CODE REVIEW

**Comment by NicolasVanBossuyt** [comment]
I think while enumerating, PCI device should be placed in /dev/pci. And userspace interacte with then through ioctl but I still need to read more about PCI to see if this a good idea

File: packages/skift.hjert.kernel/sources/pci.c [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#include "kernel/processor.h"
+
+static u16 pci_read_word(u8 bus, u8 slot, u8 func, u8 offset) {
+    u32 address;
+    u32 lbus  = (u32)bus;
+    u32 lslot = (u32)slot;
+    u32 lfunc = (u32)func;
+    u16 tmp = 0;
+    address = (u32)((lbus << 16) | (lslot << 11) | (lfunc << 8) | (offset & 0xfc) | ((u32)0x80000000));
+    outw(0xCF8, address);
```
[the remark this answers is not in the document]
Comment (SpaceboyRoss01) [reply packages/skift.hjert.kernel/sources/pci.c]: The code is based on the snippet function on the osdev wiki.

File: packages/skift.hjert.kernel/includes/kernel/pci.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#pragma once
+
+#include <skift/generic.h>
+
+typedef struct {
+} pci_dev_t;
+
+// vim:set ts=4 sw=4 et:
```
[the remark this answers is not in the document]
Comment (SpaceboyRoss01) [reply packages/skift.hjert.kernel/includes/kernel/pci.h]: Actually, it is common for people to do this when they use VIM.

**Comment by SpaceboyRoss01** [comment]
Currently errors
```
packages/skift.hjert.kernel/sources/dev/pci.c: In function 'pci_setup':
packages/skift.hjert.kernel/sources/dev/pci.c:27:5: error: implicit declaration of function 'FILESYSTEM_MKDEV' [-Werror=implicit-function-declaration]
     FILESYSTEM_MKDEV("pci", pci_device);
```

**Comment by NicolasVanBossuyt** [comment]
This because you are working on the master branch, juste checkout the develop branch and it should compile 😉

**Comment by NicolasVanBossuyt** [comment]
Maybe useful or not, but I wrote a pci driver for skiftOS, which you might want to check. 😉  https://github.com/maker-dev/core-one/commit/319c1eace4405eab813237911990f5edf13800d2

**Pull request closed without merging**